content/en/blog/_posts/2022-08-15-immutability-with-cel.md

```diff
@@ -0,0 +1,526 @@
+---
+layout: blog
+title: "Enforce CRD Immutability with CEL Transition Rules"
+date: 2022-08-15
+slug: enforce-immutability-using-cel
+canonicalUrl: https://www.kubernetes.dev/blog/2022/08/15/enforce-immutability-using-cel
+---
+
+**Author:** [Alexander Zielenski (Google)](https://github.com/alexzielenski)
+
+---
+
+# Introduction
+Immutable fields can be found in a few places in the built-in Kubernetes types such as PodSpec, EphemeralContainers, Finalizers, and more.
+
+Until recently the best way to create immutable fields for Custom Resource Definitions has been to create an admission webhook: an unfortunate tradeoff for the common case of immutable fields.
+
+Beta since Kubernetes 1.25, CEL Validation Rules allow CRD authors to express validation constraints on their fields using a rich expression language, [CEL](https://github.com/google/cel-spec). This article explores how you can use validation rules to implement a few common immutability patterns.
+# Validation Rules Basics
+The new support for CEL validation rules in Kubernetes allows CRD authors to add complicated admission logic for their resources without writing any code!
+
+For example, A CEL rule to constrain a field `maximumSize` to be greater than a `minimumSize` for a CRD might look like the following:
+```yaml
+   rule: self.maximumSize > self.minimumSize
+message: “Maximum size must be greater than minimum size.”
+```
+The rule field contains an expression written in CEL. `self` is a special keyword in CEL which refers to the object whose type contains the rule.
+
+The message field is an error message which will be sent to Kubernetes clients whenever this particular rule is not satisfied.
+## More Information
+For more details about the capabilities and limitations of Validation Rules using CEL, please refer to [the documentation](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules). The [cel-spec](https://github.com/google/cel-spec) is also a good reference for information specifically related to the language.
+
+# Immutability Patterns with CEL Validation Rules
+This section implements several common use cases for immutability in Kubernetes Custom Resource Definitions using validation rules expressed as [kubebuilder marker comments](https://book.kubebuilder.io/reference/markers/crd.html).
+
+We will also show the resultant OpenAPI generated by the kubebuilder marker comments so that those not using kubebuilder can also follow along.
```

Comment:
```suggestion
For more details about the capabilities and limitations of Validation Rules using CEL, please refer to [validation rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules). The [CEL specification](https://github.com/google/cel-spec) is also a good reference for information specifically related to the language.

## Immutability patterns with CEL validation rules

This section implements several common use cases for immutability in Kubernetes
CustomResourceDefinitions, using validation rules expressed as [kubebuilder marker comments](https://book.kubebuilder.io/reference/markers/crd.html).

I will also show the resultant OpenAPI generated by the kubebuilder marker comments so that those not using kubebuilder can also follow along.
```